cmd/agent/app/jmx.go

```diff
@@ -170,11 +181,35 @@ func runJmxCommandConsole(command string) error {
 
 	common.LoadComponents(config.Datadog.GetString("confd_path"))
 
-	err = standalone.ExecJMXCommandConsole(command, cliSelectedChecks, logLevel)
+	if discoveryRetryInterval > discoveryTimeout {
+		fmt.Println("The discovery retry interval", discoveryRetryInterval, "is higher than the discovery timeout", discoveryTimeout)
+		fmt.Println("Setting the discovery retry interval to", discoveryTimeout)
+		discoveryRetryInterval = discoveryTimeout
+	}
+
+	allConfigs := common.WaitForConfigs(time.Duration(discoveryRetryInterval)*time.Second, time.Duration(discoveryTimeout)*time.Second, selectedCheckMatcherBuilder(cliSelectedChecks, discoveryMinInstances))
+
+	err = standalone.ExecJMXCommandConsole(command, cliSelectedChecks, logLevel, allConfigs)
 
 	if runtime.GOOS == "windows" {
 		standalone.PrintWindowsUserWarning("jmx")
 	}
 
 	return err
 }
+
+// selectedCheckMatcherBuilder returns a function that returns true if the number of configs found for the
+// check name is more or equal to min instances
+func selectedCheckMatcherBuilder(checkNames []string, minInstances uint) func(configs []integration.Config) bool {
```

Comment:
As discussed on slack, this function and `checkMatcherBuilder` can be merged now.

Reply:
we'll also want to add a comment to make it explicit that we expect the function to always return `false` when `jmx list` is called with no check names (in which case the timeout would always be hit, which is desirable to maximize the chances of actually having the configs from all sources)